File: 3953-maximum-score-with-co-prime-element/3953-maximum-score-with-co-prime-element.ts
function maxScore(nums: number[], maxVal: number): number {
    let mx = maxVal;

    for (const x of nums) {
        if (x > mx) mx = x;
    }

    const freq = new Array<number>(mx + 1).fill(0);
    for (const x of nums) {
        freq[x]++;
    }

    const spf = Array.from({ length: mx + 1 }, (_, i) => i);
    const mul = new Array<number>(mx + 1).fill(0);

    for (let i = 2; i <= mx; i++) {
        if (spf[i] === i) {
            for (let j = i * 2; j <= mx; j += i) {
                if (spf[j] === j) spf[j] = i;
            }
        }

        let cnt = 0;
        for (let j = i; j <= mx; j += i) {
            cnt += freq[j];
        }
        mul[i] = cnt;
    }

    let ans = freq[1] > 0 ? 1 : 0;

    for (let i = mx; i >= 2; i--) {
        if (i <= ans) break;

        if (freq[i] === 0 && i > maxVal) continue;

        let x = i;
        const primes: number[] = [];

        while (x !== 1) {
            const p = spf[x];
            primes.push(p);
            while (x % p === 0) x = Math.floor(x / p);
        }

        let bad = -freq[i];
        const total = 1 << primes.length;

        for (let mask = 1; mask < total; mask++) {
            let prod = 1;
            let bits = 0;

            for (let k = 0; k < primes.length; k++) {
                if ((mask >> k) & 1) {
                    prod *= primes[k];
                    bits++;
                }
            }

            if (bits & 1) bad += mul[prod];
            else bad -= mul[prod];
        }

        const cost =
            freq[i] > 0
                ? bad + freq[i] - 1
                : Math.max(bad, 1);

        ans = Math.max(ans, i - cost);
    }

    return ans;
}

File: 3953-maximum-score-with-co-prime-element/3953-maximum-score-with-co-prime-element.scala
object Solution {
    def maxScore(nums: Array[Int], maxVal: Int): Int = {
        var mx = maxVal

        for (x <- nums) {
            if (x > mx) mx = x
        }

        val freq = Array.fill(mx + 1)(0)
        for (x <- nums) {
            freq(x) += 1
        }

        val spf = Array.tabulate(mx + 1)(i => i)
        val mul = Array.fill(mx + 1)(0)

        for (i <- 2 to mx) {
            if (spf(i) == i) {
                var j = i * 2
                while (j <= mx) {
                    if (spf(j) == j) {
                        spf(j) = i
                    }
                    j += i
                }
            }

            var cnt = 0
            var j = i
            while (j <= mx) {
                cnt += freq(j)
                j += i
            }
            mul(i) = cnt
        }

        var ans = if (freq(1) > 0) 1 else 0

        var i = mx
        while (i >= 2) {
            if (i <= ans) {
                return ans
            }

            if (!(freq(i) == 0 && i > maxVal)) {
                var x = i
                val primes = scala.collection.mutable.ArrayBuffer[Int]()

                while (x != 1) {
                    val p = spf(x)
                    primes += p

                    while (x % p == 0) {
                        x /= p
                    }
                }

                var bad = -freq(i)
                val total = 1 << primes.length

                var mask = 1
                while (mask < total) {
                    var prod = 1
                    var bits = 0

                    var k = 0
                    while (k < primes.length) {
                        if (((mask >> k) & 1) != 0) {
                            prod *= primes(k)
                            bits += 1
                        }
                        k += 1
                    }

                    if ((bits & 1) != 0) {
                        bad += mul(prod)
                    } else {
                        bad -= mul(prod)
                    }

                    mask += 1
                }

                val cost =
                    if (freq(i) > 0)
                        bad + freq(i) - 1
                    else
                        math.max(bad, 1)

                ans = math.max(ans, i - cost)
            }

            i -= 1
        }

        ans
    }
}